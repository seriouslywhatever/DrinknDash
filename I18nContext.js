import React, { createContext, useState, useContext, useEffect } from 'react';
import I18n from 'react-native-i18n';
import en from './translations/en.json';
import nl from './translations/nl.json';
import AsyncStorage from '@react-native-async-storage/async-storage';

I18n.fallbacks = true;
I18n.translations = { en, nl };

const I18nContext = createContext();

/**
 * Provider for changing the language of the application.
 * Based on value theme which gets changed in the Settings screen.
 * default language is english.
 */
export const I18nProvider = ({ children }) => {

    useEffect(() => {
        const checkForPreference = async () => {
            const language = await AsyncStorage.getItem('locale');
            if (language) {
                I18n.locale = language;
            } else {
                I18n.locale = 'en';
            }
        }
        checkForPreference();
    }, []);

    const [locale, setLocale] = useState(I18n.locale);

    //Changes the language of the application. 
    const changeLanguage = (newLocale) => {
        if (newLocale) {
            I18n.locale = newLocale;
            setLocale(newLocale);
        } else {
            I18n.locale = locale;
        }
    };

    return (
        <I18nContext.Provider value={{ locale, changeLanguage }}>
            {children}
        </I18nContext.Provider>
    );
};

export const useI18n = () => {
    return useContext(I18nContext);
};