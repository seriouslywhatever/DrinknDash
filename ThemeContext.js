import AsyncStorage from '@react-native-async-storage/async-storage';
import React, { createContext, useState, useContext, useEffect } from 'react';

const ThemeContext = createContext();

/**
 * Provider to change the theme of the application. 
 * Based on value theme which gets changed in the Settings screen.
 */
export const ThemeProvider = ({ children }) => {

    //Values for light mode of the application
    const lightTheme = {
        backgroundImage: require('./assets/wood.jpg'),
        trackImage: require('./assets/track.jpg'),
        fenceImage: require('./assets/fence.jpg'),
        color: '#343434', //dark
        backgroundColor: '#EDEADE', //light
    }


    //Values for Dark mode of the application
    const darkTheme = {
        backgroundImage: require('./assets/wood_dark.jpg'),
        trackImage: require('./assets/track_dark.jpg'),
        fenceImage: require('./assets/fence_dark.jpg'),
        color: '#EDEADE', //light
        backgroundColor: '#343434', //dark
    }

    const [theme, setTheme] = useState(lightTheme);

    //Method to change the theme of the application.
    const changeTheme = (value) => {
        console.log(value);
        setTheme(value ? lightTheme : darkTheme);
    };

    useEffect(() => {
        const checkForPreference = async () => {
            const savedTheme = await AsyncStorage.getItem('theme');
            if (savedTheme) {
                changeTheme(JSON.parse(savedTheme));
            }
        }
        checkForPreference();
    }, []);

    return (
        <ThemeContext.Provider value={{ theme, changeTheme }}>
            {children}
        </ThemeContext.Provider>
    );
};

export const useTheme = () => {
    return useContext(ThemeContext);
};
