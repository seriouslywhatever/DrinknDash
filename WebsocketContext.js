import React, { createContext, useContext, useState, useEffect } from 'react';

const WebSocketContext = createContext();

/**
 * Provider for socket to communicate with server.
 */
export const WebSocketProvider = ({ children }) => {
    const [socket, setSocket] = useState(null);

    useEffect(() => {
        const ws = new WebSocket('ws://192.168.2.2:8080'); //ip of wifi 

        ws.onopen = () => {
            console.log('WebSocket connection opened');
        };

        ws.onclose = () => {
            console.log('WebSocket connection closed!');
        };

        setSocket(ws);
        return () => {
            ws.close();
        };
    }, []);

    return (
        <WebSocketContext.Provider value={{ socket }}>
            {children}
        </WebSocketContext.Provider>
    );
};

export const useWebSocket = () => {
    return useContext(WebSocketContext);
};