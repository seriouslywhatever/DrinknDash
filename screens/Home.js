import { useCallback, useEffect, useState } from 'react';
import { View, Text, TouchableOpacity, FlatList, Image, ImageBackground, StyleSheet, RefreshControl, SafeAreaView, ScrollView } from 'react-native';
import { useWebSocket } from '../WebsocketContext';
import AsyncStorage from '@react-native-async-storage/async-storage';
import { useTranslation } from 'react-i18next';
import { useTheme } from '../ThemeContext';

export default Home = ({ navigation }) => {

    const { socket } = useWebSocket();
    const [username, setUsername] = useState(" ");
    const [list, setList] = useState([]);
    const [lobbyId, setLobbyId] = useState();

    const { theme } = useTheme(); //Provider of style values based on application theme.
    const { t } = useTranslation();
    const [refreshing, setRefreshing] = useState(false);

    //Refresh list of lobbies on pull down. 
    const onRefresh = useCallback(() => {
        setRefreshing(true);
        try {
            socket.send("LIST");
        } catch (error) {
            console.error(error);
        }
        setTimeout(() => {
            setRefreshing(false);
        }, 2000);
    }, []);

    useEffect(() => {
        if (socket) {
            socket.onmessage = (event) => {

                // Handle incoming messages from the server
                var server_message = event.data.split(' ');
                if (server_message[0] == "LIST") {
                    if (server_message[1] != "undefined") {
                        const list = event.data.substring(event.data.indexOf(' ') + 1);
                        console.log(list);
                        setList(JSON.parse(list));
                    }
                } else if (server_message[0] == "LOBBY") {
                    console.log(server_message);
                    if (server_message[1]) {
                        navigation.navigate("Lobby", { id: server_message[1] });
                    } else {
                        navigation.navigate("Lobby", { id: lobbyId });
                    }
                }
            };
        }

        //Ensures username retrieved which is used for game creation 
        retrieveUser = async () => {
            try {
                const usernameHolder = await AsyncStorage.getItem('username');
                if (usernameHolder) {
                    setUsername(usernameHolder);
                }
            } catch (error) {
                console.error(error);
            }
        }
        //populate list of lobbies on launch.
        retrieveLobbies = async () => {
            try {
                socket.send("LIST");
            } catch (error) {
                console.error(error);
            }
        }
        retrieveUser();
        retrieveLobbies();

    }, [lobbyId]);

    const renderItem = useCallback(({ item }) => (
        <TouchableOpacity
            onPress={() => {
                socket.send("JOIN " + JSON.stringify(item.id));
                setLobbyId(item.id);
            }}
            style={[styles.itemContainer, { backgroundColor: theme.backgroundColor }]}>
            <View style={styles.itemTitleContainer}>
                <Text numberOfLines={1} style={[styles.itemTitleText, { color: theme.color }]}>{item.title}</Text>
            </View>
            <View style={styles.playersContainer}>
                <View>
                    <Text style={[styles.playerCountText, { color: theme.color }]}>{item.total_player}</Text>
                </View>
                <Text style={{ color: theme.color }}>{t('playersText')}</Text>
            </View>
        </TouchableOpacity>
    ), []);

    return (
        <SafeAreaView style={{ flex: 1 }}>
            <View style={{ flex: 1, backgroundColor: theme.backgroundColor }}>
                <View style={{ flex: 1 }}>
                    <ImageBackground source={theme.backgroundImage} style={styles.backgroundImage}>
                        <TouchableOpacity
                            onPress={() => navigation.navigate("Settings")}
                            style={styles.settingsButton}>
                            <Image source={require('../assets/settings.jpg')} style={styles.settingsImage} />
                        </TouchableOpacity>
                        <TouchableOpacity
                            onPress={async () => {
                                try {
                                    const usernameHolder = await AsyncStorage.getItem('username');
                                    if (usernameHolder) {
                                        await AsyncStorage.removeItem('username');
                                        socket.send("LOGOUT");
                                        navigation.replace("Login");
                                    }
                                } catch (error) {
                                    console.error('Error storing credentials:', error);
                                }
                            }}
                            style={styles.logoutButton}>
                            <Text style={{ color: theme.color }}>{t('logoutText')}</Text>
                        </TouchableOpacity>
                        <Text style={[styles.headerText, { color: theme.color }]}>{t('lobbiesText')}</Text>
                        <Text style={{ color: theme.color }}>{username}</Text>
                    </ImageBackground>
                </View>
                <View style={styles.hr} />
                <View style={{ flex: 6 }}>
                    <FlatList
                        data={list}
                        renderItem={renderItem}
                        keyExtractor={(item) => item.id}
                        refreshControl={
                            <RefreshControl refreshing={refreshing} onRefresh={onRefresh} />
                        }
                    />
                </View>
                <View style={styles.buttonContainer}>
                    <TouchableOpacity
                        onPress={() => {
                            if (socket && socket.readyState === WebSocket.OPEN) {
                                socket.send("CREATE " + JSON.stringify({ username: username }));
                            }
                        }}
                        style={[styles.button, { backgroundColor: theme.backgroundColor }]}>
                        <Text style={[styles.buttonText, { color: theme.color }]}>{t('createLobbyText')}</Text>
                    </TouchableOpacity>
                </View >
            </View>
        </SafeAreaView >
    )
};

const styles = StyleSheet.create({
    backgroundImage: {
        flex: 1,
        justifyContent: 'center',
        alignItems: 'center'
    },
    settingsButton: {
        position: 'absolute',
        top: '5%',
        left: '2%'
    },
    settingsImage: {
        height: 30,
        width: 30
    },
    logoutButton: {
        position: 'absolute',
        top: '5%',
        right: '2%'
    },
    headerText: {
        fontSize: 32,
        fontWeight: '500',
    },
    hr: {
        borderWidth: 2,
        width: '100%'
    },
    buttonContainer: {
        position: 'absolute',
        bottom: '5%',
        right: '5%'
    },
    button: {
        borderRadius: 50,
        padding: 20,
        opacity: 0.8,
        borderWidth: 1
    },
    buttonText: {
        textAlign: 'center', fontSize: 16
    },
    itemTitleContainer: {
        flex: 6,
        justifyContent: 'center',
        marginStart: '5%'
    },
    itemContainer: {
        flexDirection: 'row',
        flex: 1,
        borderWidth: 1,
    },
    itemTitleText: {
        fontSize: 22,
        fontWeight: '500',
    },
    playersContainer: {
        flex: 2,
        flexDirection: 'column',
        justifyContent: 'center',
        margin: '5%',
        alignItems: 'center'
    },
    playerCountText: {
        fontSize: 18,
        textAlign: 'center',
        padding: 5,
        borderWidth: 1
    }
});