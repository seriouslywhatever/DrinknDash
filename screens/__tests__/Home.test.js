import React from 'react';
import { render, fireEvent, waitFor } from '@testing-library/react-native';
import renderer from 'react-test-renderer';
import { ThemeProvider } from '../../ThemeContext';
import { WebSocketProvider } from '../../WebsocketContext';
import { I18nProvider } from '../../I18nContext';
import  Home from '../Home'

jest.mock('../../ThemeContext', () => ({
    useTheme: () => ({ theme: { color: 'black', backgroundImage: require('../../assets/wood.jpg') } })
}));

jest.mock('../../WebsocketContext', () => ({
    useWebSocket: () => ({ socket: { send: jest.fn() } })
}));

jest.mock('../../i18nContext', () => ({
    useI18n: jest.fn().mockReturnValue({
        locale: 'en',
        changeLanguage: jest.fn(),
    }),
}));

describe('Home', () => {

    test('renders correctly', () => {
        const tree = renderer.create(<Home />);
        expect(tree).toMatchSnapshot();
    });

});