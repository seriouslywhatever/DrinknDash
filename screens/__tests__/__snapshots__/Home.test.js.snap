// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home renders correctly 1`] = `
<RCTSafeAreaView
  style={
    {
      "flex": 1,
    }
  }
>
  <View
    style={
      {
        "backgroundColor": undefined,
        "flex": 1,
      }
    }
  >
    <View
      style={
        {
          "flex": 1,
        }
      }
    >
      <View
        accessibilityIgnoresInvertColors={true}
        style={
          {
            "alignItems": "center",
            "flex": 1,
            "justifyContent": "center",
          }
        }
      >
        <Image
          source={1}
          style={
            [
              {
                "bottom": 0,
                "left": 0,
                "position": "absolute",
                "right": 0,
                "top": 0,
              },
              {
                "height": undefined,
                "width": undefined,
              },
              undefined,
            ]
          }
        />
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "left": "2%",
              "opacity": 1,
              "position": "absolute",
              "top": "5%",
            }
          }
        >
          <Image
            source={1}
            style={
              {
                "height": 30,
                "width": 30,
              }
            }
          />
        </View>
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "opacity": 1,
              "position": "absolute",
              "right": "2%",
              "top": "5%",
            }
          }
        >
          <Text
            style={
              {
                "color": "black",
              }
            }
          >
            [missing "en.logoutText" translation]
          </Text>
        </View>
        <Text
          style={
            [
              {
                "fontSize": 32,
                "fontWeight": "500",
              },
              {
                "color": "black",
              },
            ]
          }
        >
          [missing "en.lobbiesText" translation]
        </Text>
        <Text
          style={
            {
              "color": "black",
            }
          }
        >
           
        </Text>
      </View>
    </View>
    <View
      style={
        {
          "borderWidth": 2,
          "width": "100%",
        }
      }
    />
    <View
      style={
        {
          "flex": 6,
        }
      }
    >
      <RCTScrollView
        data={[]}
        getItem={[Function]}
        getItemCount={[Function]}
        keyExtractor={[Function]}
        onContentSizeChange={[Function]}
        onLayout={[Function]}
        onMomentumScrollBegin={[Function]}
        onMomentumScrollEnd={[Function]}
        onScroll={[Function]}
        onScrollBeginDrag={[Function]}
        onScrollEndDrag={[Function]}
        refreshControl={
          <RefreshControlMock
            onRefresh={[Function]}
            refreshing={false}
          />
        }
        removeClippedSubviews={false}
        renderItem={[Function]}
        scrollEventThrottle={50}
        stickyHeaderIndices={[]}
        viewabilityConfigCallbackPairs={[]}
      >
        <RCTRefreshControl />
        <View />
      </RCTScrollView>
    </View>
    <View
      style={
        {
          "bottom": "5%",
          "position": "absolute",
          "right": "5%",
        }
      }
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "backgroundColor": undefined,
            "borderRadius": 50,
            "borderWidth": 1,
            "opacity": 0.8,
            "padding": 20,
          }
        }
      >
        <Text
          style={
            [
              {
                "fontSize": 16,
                "textAlign": "center",
              },
              {
                "color": "black",
              },
            ]
          }
        >
          [missing "en.createLobbyText" translation]
        </Text>
      </View>
    </View>
  </View>
</RCTSafeAreaView>
`;
